Proposal for issue #31

AxelDelmas opened the pull request:
Here's a fix proposal for issue #31, to be used as a basis for discussion (need a little bit of cleaning if this approach is accepted)

mangui commented:
thanks for the PR @AxelDelmas 
this change https://github.com/AxelDelmas/hls.js/commit/62189fa01f3ac018f7b0de081786033c32852bf8 makes sense.
regarding https://github.com/AxelDelmas/hls.js/commit/f93a68cce65a1d705a5e67fb5fdb12969426d5e2, I understand your point as well, but for consistency purpose, i think it would be better to express it in multiple of TARGET DURATION instead, to be in sync with https://github.com/dailymotion/hls.js/blob/master/API.md#livesyncdurationcount

it could be liveMaxLatencyDurationCount or something similar for example

AxelDelmas commented:
Ok I see. I didn't really pay attention to the fact that `liveSyncDurationCount` was configurable (I mentionned it as a possible improvement in issue #31 but in fact it's already there).

The problem with that stream is that even though the target chunk duration is set to 2s in the media server configuration, in practice `EXT-X-TARGETDURATION` varies between each manifest reload. I think Wowza enforces that every segment begins with a keyframe (which is actually good for playback in chrome), so that's why the fragment duration is variable: sometimes 10s or more. It must then set `EXT-X-TARGETDURATION` to the rounded average fragment length in the current index, instead of setting it to 2s (which would be completely irrelevant in most cases with this playlist).
This is probably more a packager-side issue, but reasoning in seconds with `liveSyncDuration*` and `liveMaxLatencyDuration*` might help with this kind of inconsistencies.

mangui commented:
@AxelDelmas I did a bunch of changes on buffer-controller to handle playlist drift and sliding, 
start position is now re-evaluated automatically.

regarding your second point : defining a max latency for live playlist, in case this max is reached, seek back to live point : 
it would be quite straightforward to add here : https://github.com/dailymotion/hls.js/blob/master/src/controller/buffer-controller.js#L215

instead of checking that end of buffer is located before the start of the sliding playlist :

`if (bufferEnd < start)`

we would need to check

`if (bufferEnd < Math.max(start,end-this.config.liveMaxLatencyDurationCount*levelDetails.targetduration))`

that would ensure that we are matching latency requirements.
the reason why i prefer to keep multiple of target duration is that this approach is more generic and could cope with any live playlists and any fragment duration change.

AxelDelmas commented:
Ok thanks a lot for the change @mangui.
I'll update my PR for the live max latency. What do you think would be a good default value for `liveMaxLatencyDurationCount` ?

mangui commented:
@AxelDelmas 
in order to avoid changing the current behavior, in which the max live latency is the start of the sliding window, i would suggest to put undefined as default value :  in case live max latency is undefined, then live max latency should match with the start of the sliding window, as it is today.

AxelDelmas commented:
Updated PR.
Implemented `liveMaxLatencyDurationCount` and added doc on new param.
I also added a fail fast if this new new param is inferior or equal to `liveSyncDurationCount` because this configuration would cause the player to seek unnecessarily all the time.

mangui commented on src/hls.js:
```diff
@@ -41,6 +41,7 @@ class Hls {
       maxBufferLength: 30,
       maxBufferSize: 60 * 1000 * 1000,
       liveSyncDurationCount:3,
+      liveMaxLatencyDurationCount: undefined, //useless, but here as a reminder that this conf parameter exists
```
setting it to Infinity is better, as Math.max(start,undefined) will return NaN

AxelDelmas commented:
That's right! Updated PR.

mangui commented:
thanks for the PR, merged
